File: app/src/main/java/com/kjjejones42/audiobookplayer/display/ListItem.java
package com.kjjejones42.audiobookplayer.display;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.kjjejones42.audiobookplayer.AudioBook;

import java.util.HashMap;
import java.util.Map;

abstract class ListItem {

    final static int TYPE_HEADING = 0;
    final static int TYPE_ITEM = 1;

    final private static Map<String, Long> idMap = new HashMap<>();

    static long getId(String name) {
        long id;
        if (idMap.containsKey(name)) {
            Long value = idMap.get(name);
            if (value == null) {
                throw new RuntimeException();
            }
            id = value;
        } else {
            id = idMap.entrySet().size();
            idMap.put(name, id);
        }
        return id;
    }

    public abstract long getId();

    public abstract int getCategory();

    public abstract long getTimeStamp();

    public abstract int getHeadingOrItem();

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj instanceof ListItem) {
            return obj.toString().equals(this.toString());
        }
        return super.equals(obj);
    }

    public static class AudioBookContainer extends ListItem {
        final AudioBook book;
        final private long id;

        AudioBookContainer(AudioBook book) {
            this.book = book;
            this.id = getId(book.getUniqueId());
        }

        @Override
        public long getId() {
            return id;
        }

        @Override
        public int getCategory() {
            return book.getStatus();
        }

        @Override
        public int getHeadingOrItem() {
            return TYPE_ITEM;
        }


        @Override
        public long getTimeStamp() {
            return book.lastSavedTimestamp;
        }

        @NonNull
        @Override
        public String toString() {
            return book.displayName;
        }
    }

    public static class Heading extends ListItem {
        final private int category;
        final private long id;

        Heading(int title) {
            this.category = title;
            this.id = getId(getHeadingTitle());
        }

        @Override
        public long getId() {
            return id;
        }

        @Override
        public int getCategory() {
            return category;
        }

        @Override
        public int getHeadingOrItem() {
            return TYPE_HEADING;
        }

        @Override
        public long getTimeStamp() {
            return 0;
        }

        String getHeadingTitle() {
            return AudioBook.getStatusMap().get(category);
        }

        @NonNull
        @Override
        public String toString() {
            return getHeadingTitle();
        }

    }
}

File: app/src/main/java/com/kjjejones42/audiobookplayer/display/ListItem.kt
package com.kjjejones42.audiobookplayer.display

import com.kjjejones42.audiobookplayer.AudioBook
import com.kjjejones42.audiobookplayer.AudioBookStatus

enum class ListItemType(val value: Int) {
    HEADING(0),
    ITEM(1)
}

abstract class ListItem {

    companion object {
        private val idMap: MutableMap<String?, Long> = HashMap()

        fun getId(name: String?): Long {
            return idMap.computeIfAbsent(name) { idMap.entries.size.toLong() }
        }
    }

    abstract val id: Long
    abstract val category: Int
    abstract val timeStamp: Long
    abstract val type: ListItemType

    override fun equals(other: Any?): Boolean {
        if (other is ListItem) {
            return other.toString() == this.toString()
        }
        return super.equals(other)
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }

    class AudioBookContainer internal constructor(val book: AudioBook) : ListItem() {
        override val id: Long = getId(book.uniqueId)
        override val category: Int = book.status
        override val timeStamp: Long = book.lastSavedTimestamp
        override val type: ListItemType = ListItemType.ITEM

        override fun toString(): String {
            return book.displayName
        }
    }

    class Heading internal constructor(override val category: Int) : ListItem() {
        val headingTitle: String? = AudioBookStatus.entries[category].displayName
        override val timeStamp: Long = 0
        override val type: ListItemType = ListItemType.HEADING
        override val id: Long = getId(headingTitle)

        override fun toString(): String {
            return headingTitle ?: ""
        }
    }
}
